### ISSUE TITLE
Cannot authenticate with HA

### ISSUE THREAD

**Issue by AndreKR** [home-assistant/android#2729]
**Home Assistant Android version:** 2022.6.0-full

**Android version:** 10

**Phone model:** Oneplus Nord N10 (BE2029)

**Home Assistant version:** 2022.3.7

**Description of problem:**

I cannot authenticate the app with my Home Assistant.

I entered the address of my Home Assistant as external URL. (`https://mydomain.com/`). When I open the same address in a browser on my PC (or in Chrome on the phone for that matter) I get redirected to `https://mydomain.com/lovelace` and from there to `https://mydomain.com/auth/authorize` and I can login and everything works.

But when I open the app, it will just open a Chrome tab with `https://mydomain.com/lovelace/?external_auth=1` which looks like this:

![image](https://user-images.githubusercontent.com/1188538/181795639-79ae3780-d127-4557-a488-1d80081be7a6.png)

Note that that's an actual Chrome tab, in the Chrome app. It is not embedded into the Home Assistant app in any way that I could see. It shows up among my other open tabs in the Chrome app.

When I inspect that Chrome tab via Chrome remote debugging it will show a console error:

```
Uncaught (in promise) Error: External auth requires either externalApp or webkit defined on Window object.
```

I captured a logcat for the Home Assistant app's PID from the start of the Home Assistant app until the Chrome window appears (takes about 2 seconds). This is filtered by "13096", the PID of the companion app.

```
2022-07-29 17:32:10.380 1423-3026/? I/ActivityTaskManager: The Process io.homeassistant.companion.android Already Exists in BG. So sending its PID: 13096
2022-07-29 17:32:10.382 13096-13096/? W/ActivityThread: handleWindowVisibility: no activity for token android.os.BinderProxy@77686a1
2022-07-29 17:32:10.394 1423-2664/? D/Foreground_io: TOP_APP is ProcessRecord{dff54 13096:io.homeassistant.companion.android/u0a250} uid is 10250
2022-07-29 17:32:10.422 13096-13096/? D/WindowManager: Add to mViews: DecorView@7b93168[LaunchActivity], this = android.view.WindowManagerGlobal@cdeb8e0,pkg= io.homeassistant.companion.android
2022-07-29 17:32:10.423 13096-13096/? V/ViewRootImpl: The specified message queue synchronization  barrier token has not been posted or has already been removed
2022-07-29 17:32:10.453 13096-13096/? W/Choreographer: Already have a pending vsync event.  There should only be one at a time.
2022-07-29 17:32:10.456 13096-13096/? D/UrlRepository: localUrl is: true and usesInternalSsid is: false
2022-07-29 17:32:10.456 13096-13096/? D/UrlRepository: Using external URL
2022-07-29 17:32:10.457 13096-13096/? D/UrlRepository: localUrl is: true and usesInternalSsid is: false
2022-07-29 17:32:10.463 13096-14625/? D/UrlRepository: localUrl is: true and usesInternalSsid is: false
2022-07-29 17:32:10.465 1423-2270/? I/ActivityTaskManager: START u0 {cmp=io.homeassistant.companion.android/.webview.WebViewActivity (has extras)} from uid 10250 pid 13096
2022-07-29 17:32:10.485 1423-2270/? I/ActivityTaskManager: The Process io.homeassistant.companion.android Already Exists in BG. So sending its PID: 13096
2022-07-29 17:32:10.497 13096-13096/? W/ActivityThread: handleWindowVisibility: no activity for token android.os.BinderProxy@c77f26c
2022-07-29 17:32:10.509 1423-8299/? D/ConnectivityService: requestNetwork for uid/pid:10250/13096 NetworkRequest [ TRACK_DEFAULT id=12664, [ Capabilities: INTERNET&NOT_RESTRICTED&TRUSTED Uid: 10250] ]
2022-07-29 17:32:10.567 13096-14625/? E/LaunchPresenter: Issue updating Registration
    io.homeassistant.companion.android.common.data.integration.IntegrationException: Error calling integration request update_registration
        at io.homeassistant.companion.android.common.data.integration.impl.IntegrationRepositoryImpl.updateRegistration(IntegrationRepositoryImpl.kt:136)
        at io.homeassistant.companion.android.common.data.integration.impl.IntegrationRepositoryImpl$updateRegistration$1.invokeSuspend(Unknown Source:15)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
        at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:749)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
2022-07-29 17:32:10.574 13096-13096/? D/PhoneWindow: setNavigationBarColor: fffafafa
2022-07-29 17:32:10.578 13096-13096/? I/System.out: recordForce value 6
2022-07-29 17:32:10.578 13096-13096/? D/WebviewActivity: Current webview package com.google.android.webview and version 101.0.4951.61
2022-07-29 17:32:10.678 1423-8299/? D/ConnectivityService: requestNetwork for uid/pid:10250/13096 NetworkRequest [ TRACK_DEFAULT id=12666, [ Capabilities: INTERNET&NOT_RESTRICTED&TRUSTED Uid: 10250] ]
2022-07-29 17:32:10.804 13096-13096/? D/WindowManager: Add to mViews: DecorView@2d7b219[WebViewActivity], this = android.view.WindowManagerGlobal@cdeb8e0,pkg= io.homeassistant.companion.android
2022-07-29 17:32:10.807 13096-13096/? V/ViewRootImpl: The specified message queue synchronization  barrier token has not been posted or has already been removed
2022-07-29 17:32:10.821 13096-13096/? D/OpScreenModeManager: setRefreshRate view android.view.SurfaceView{9e00eea G.E...... ......I. 0,0-0,0} viewRate 0 width 0 height 0
2022-07-29 17:32:10.830 13096-16203/? D/DecorView: onWindowFocusChangedFromViewRoot hasFocus: true, DecorView@2d7b219[WebViewActivity]
2022-07-29 17:32:10.831 13096-16203/? D/ViewRootImpl[WebViewActivity]: windowFocusChanged hasFocus=true inTouchMode=true
2022-07-29 17:32:10.843 13096-13096/? W/Choreographer: Already have a pending vsync event.  There should only be one at a time.
2022-07-29 17:32:10.883 13096-13096/? D/UrlRepository: localUrl is: true and usesInternalSsid is: false
2022-07-29 17:32:10.883 13096-13096/? D/UrlRepository: Using external URL
2022-07-29 17:32:10.885 13096-14625/? D/UrlRepository: localUrl is: true and usesInternalSsid is: false
2022-07-29 17:32:10.886 13096-14626/? I/WM-WorkerWrapper: Worker result SUCCESS for Work [ id=122ae4c0-15a9-4151-b1f0-46bf51c915a4, tags={ io.homeassistant.companion.android.websocket.WebsocketManager } ]
2022-07-29 17:32:10.892 13096-14960/? D/UrlRepository: localUrl is: true and usesInternalSsid is: false
2022-07-29 17:32:10.894 13096-14960/? D/UrlRepository: localUrl is: true and usesInternalSsid is: false
2022-07-29 17:32:10.894 13096-14960/? D/UrlRepository: Using external URL
2022-07-29 17:32:10.896 13096-14960/? D/UrlRepository: localUrl is: true and usesInternalSsid is: false
2022-07-29 17:32:10.896 13096-14960/? D/UrlRepository: Using external URL
2022-07-29 17:32:10.927 13096-13096/? D/WebviewActivity: Launching browser
2022-07-29 17:32:10.942 13096-16762/? W/okhttp.OkHttpClient: A connection to https://[REDACTED]/ was leaked. Did you forget to close a response body? To see where this was allocated, set the OkHttpClient logger level to FINE: Logger.getLogger(OkHttpClient.class.getName()).setLevel(Level.FINE);
2022-07-29 17:32:10.948 1423-8724/? I/ActivityTaskManager: START u0 {act=android.intent.action.VIEW dat=https://[REDACTED]/... cmp=com.android.chrome/com.google.android.apps.chrome.IntentDispatcher} from uid 10250 pid 13096
2022-07-29 17:32:10.991 13096-15613/? D/ViewRootImpl[WebViewActivity]: windowFocusChanged hasFocus=false inTouchMode=true
2022-07-29 17:32:11.026 13096-14960/? E/IntegrationRepository: Issue getting new version from core.
    io.homeassistant.companion.android.common.data.authentication.AuthorizationException
        at io.homeassistant.companion.android.common.data.authentication.impl.AuthenticationRepositoryImpl.ensureValidSession(AuthenticationRepositoryImpl.kt:231)
        at io.homeassistant.companion.android.common.data.authentication.impl.AuthenticationRepositoryImpl.access$ensureValidSession(AuthenticationRepositoryImpl.kt:27)
        at io.homeassistant.companion.android.common.data.authentication.impl.AuthenticationRepositoryImpl$ensureValidSession$1.invokeSuspend(Unknown Source:15)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
        at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:749)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
2022-07-29 17:32:11.026 13096-14960/? D/SensorWorker: Updating all Sensors.
2022-07-29 17:32:11.051 13096-14639/? I/WM-WorkerWrapper: Work [ id=0b1fd8a4-db46-426a-8fc8-93887d409a2f, tags={ io.homeassistant.companion.android.sensors.SensorWorker } ] was cancelled
    java.util.concurrent.CancellationException: Task was cancelled.
        at androidx.work.impl.utils.futures.AbstractFuture.cancellationExceptionWithCause(AbstractFuture.java:1184)
        at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:514)
        at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:475)
        at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:311)
        at androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
2022-07-29 17:32:11.066 13096-13096/? I/WM-SystemFgDispatcher: Stopping foreground service
2022-07-29 17:32:11.088 13096-14623/? D/UrlRepository: localUrl is: true and usesInternalSsid is: false
2022-07-29 17:32:11.089 13096-14623/? D/UrlRepository: localUrl is: true and usesInternalSsid is: false
2022-07-29 17:32:11.089 13096-14623/? D/UrlRepository: Using external URL
2022-07-29 17:32:11.091 13096-14623/? D/UrlRepository: localUrl is: true and usesInternalSsid is: false
2022-07-29 17:32:11.091 13096-14623/? D/UrlRepository: Using external URL
2022-07-29 17:32:11.098 13096-14640/? E/WM-WorkerWrapper: Didn't find WorkSpec for id 0b1fd8a4-db46-426a-8fc8-93887d409a2f
2022-07-29 17:32:11.208 13096-16816/? D/WebSocketRepository: Websocket: onOpen
2022-07-29 17:32:11.209 13096-16816/? D/WebSocketRepository: Websocket: onMessage (text)
2022-07-29 17:32:11.210 13096-16816/? D/WebSocketRepository: Message number null received: {"type":"auth_required","ha_version":"2022.3.7"}
2022-07-29 17:32:11.213 13096-14623/? D/WebSocketRepository: Auth Requested
2022-07-29 17:32:11.217 13096-14623/? E/IntegrationRepository: Issue getting new version from core.
    io.homeassistant.companion.android.common.data.authentication.AuthorizationException
        at io.homeassistant.companion.android.common.data.authentication.impl.AuthenticationRepositoryImpl.ensureValidSession(AuthenticationRepositoryImpl.kt:231)
        at io.homeassistant.companion.android.common.data.authentication.impl.AuthenticationRepositoryImpl.access$ensureValidSession(AuthenticationRepositoryImpl.kt:27)
        at io.homeassistant.companion.android.common.data.authentication.impl.AuthenticationRepositoryImpl$ensureValidSession$1.invokeSuspend(Unknown Source:15)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
        at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:749)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
2022-07-29 17:32:11.217 13096-14623/? D/SensorWorker: Updating all Sensors.
2022-07-29 17:32:11.220 13096-14621/? I/WM-Processor: Moving WorkSpec (65f820e6-bb0d-4604-ab59-c9fe1c00083f) to the foreground
2022-07-29 17:32:11.270 13096-13096/? I/WM-SystemFgDispatcher: Started foreground service Intent { act=ACTION_START_FOREGROUND cmp=io.homeassistant.companion.android/androidx.work.impl.foreground.SystemForegroundService (has extras) }
2022-07-29 17:32:11.286 13096-14623/? D/UrlRepository: localUrl is: true and usesInternalSsid is: false
2022-07-29 17:32:11.287 13096-14623/? D/UrlRepository: Using external URL
2022-07-29 17:32:11.288 13096-14623/? D/UrlRepository: localUrl is: true and usesInternalSsid is: false
2022-07-29 17:32:11.288 13096-14623/? D/UrlRepository: Using external URL
2022-07-29 17:32:11.384 13096-16815/? D/WebSocketRepository: Websocket: onOpen
2022-07-29 17:32:11.385 13096-16815/? D/WebSocketRepository: Websocket: onMessage (text)
2022-07-29 17:32:11.386 13096-16815/? D/WebSocketRepository: Message number null received: {"type":"auth_required","ha_version":"2022.3.7"}
2022-07-29 17:32:11.386 13096-14623/? D/WebSocketRepository: Auth Requested
2022-07-29 17:32:11.398 13096-14625/? E/IntegrationRepository: Issue getting new version from core.
    io.homeassistant.companion.android.common.data.authentication.AuthorizationException
        at io.homeassistant.companion.android.common.data.authentication.impl.AuthenticationRepositoryImpl.ensureValidSession(AuthenticationRepositoryImpl.kt:231)
        at io.homeassistant.companion.android.common.data.authentication.impl.AuthenticationRepositoryImpl.access$ensureValidSession(AuthenticationRepositoryImpl.kt:27)
        at io.homeassistant.companion.android.common.data.authentication.impl.AuthenticationRepositoryImpl$ensureValidSession$1.invokeSuspend(Unknown Source:15)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
        at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:749)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
2022-07-29 17:32:11.399 13096-14625/? D/UrlRepository: localUrl is: true and usesInternalSsid is: false
2022-07-29 17:32:11.400 13096-14625/? D/UrlRepository: localUrl is: true and usesInternalSsid is: false
2022-07-29 17:32:11.400 13096-14625/? D/UrlRepository: Using external URL
2022-07-29 17:32:11.403 13096-14625/? D/UrlRepository: localUrl is: true and usesInternalSsid is: false
2022-07-29 17:32:11.404 13096-14625/? D/UrlRepository: Using external URL
2022-07-29 17:32:11.535 13096-14625/? E/IntegrationRepository: Issue getting new version from core.
    io.homeassistant.companion.android.common.data.authentication.AuthorizationException
        at io.homeassistant.companion.android.common.data.authentication.impl.AuthenticationRepositoryImpl.ensureValidSession(AuthenticationRepositoryImpl.kt:231)
        at io.homeassistant.companion.android.common.data.authentication.impl.AuthenticationRepositoryImpl.access$ensureValidSession(AuthenticationRepositoryImpl.kt:27)
        at io.homeassistant.companion.android.common.data.authentication.impl.AuthenticationRepositoryImpl$ensureValidSession$1.invokeSuspend(Unknown Source:15)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
        at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:749)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
2022-07-29 17:32:11.566 13096-16813/? D/WebSocketRepository: Websocket: onOpen
2022-07-29 17:32:11.566 13096-16813/? D/WebSocketRepository: Websocket: onMessage (text)
2022-07-29 17:32:11.567 13096-16813/? D/WebSocketRepository: Message number null received: {"type":"auth_required","ha_version":"2022.3.7"}
2022-07-29 17:32:11.567 13096-14623/? D/WebSocketRepository: Auth Requested
2022-07-29 17:32:11.667 13096-16814/? D/WebSocketRepository: Websocket: onOpen
2022-07-29 17:32:11.669 13096-16814/? D/WebSocketRepository: Websocket: onMessage (text)
2022-07-29 17:32:11.670 13096-16814/? D/WebSocketRepository: Message number null received: {"type":"auth_required","ha_version":"2022.3.7"}
2022-07-29 17:32:11.671 13096-14623/? D/WebSocketRepository: Auth Requested
2022-07-29 17:32:11.694 13096-14625/? D/UrlRepository: localUrl is: true and usesInternalSsid is: false
2022-07-29 17:32:11.696 13096-14625/? D/UrlRepository: localUrl is: true and usesInternalSsid is: false
2022-07-29 17:32:11.696 13096-14625/? D/UrlRepository: Using external URL
2022-07-29 17:32:11.698 13096-14625/? D/UrlRepository: localUrl is: true and usesInternalSsid is: false
2022-07-29 17:32:11.698 13096-14625/? D/UrlRepository: Using external URL
```

**Additional information:**

My Home Assistant is behind a reverse proxy (Caddy) and has a Let's encrypt certificate.

This setup has worked before, with a different phone (Android 5 and possibly an older version of the companion app). Nothing has changed on the Home Assistant side, I just got a new phone.

**Comment by dshokouhi** [comment]
Can you try these steps: https://companion.home-assistant.io/docs/troubleshooting/faqs#starting-fresh-with-the-android-app

**Comment by dshokouhi** [comment]
Also bypass your proxy and test again to see if its a proxy issue

**Comment by AndreKR** [comment]
I didn't know that the WebView can be updated separately from the system. I did that, but that alone didn't fix the issue. Then I deleted the apps storage (I had done that before though without success) and removed the phone in HA under Integrations.

After that I started fresh by entering the _external_ URL as URL (the previous attempts I had always started with the internal URL) and this time it worked.

So, no idea which of these it was, but thanks.
